File: mediashell/screens/iptv.py
"""IPTV channel list screen with group navigation"""

import pygame
from screens import BaseScreen

ACCENT  = (100, 120, 220)
TILE_BG = (28, 28, 28)
TILE_HL = (55, 55, 80)
WHITE   = (255, 255, 255)
GRAY    = (140, 140, 140)

ITEM_H  = 56
MARGIN_X = 60
MARGIN_Y = 140
LIST_W   = 860
GROUP_W  = 280


class IptvScreen(BaseScreen):
    def __init__(self, shell):
        super().__init__(shell)
        self.groups: list[str] = []
        self.channels: list[dict] = []
        self.group_idx = 0
        self.ch_idx = 0
        self.pane = "groups"  # "groups" | "channels"
        self._font_hdr  = None
        self._font_item = None
        self._font_sub  = None

    def _ensure_fonts(self) -> None:
        if self._font_hdr:
            return
        pygame.font.init()
        self._font_hdr  = pygame.font.SysFont("DejaVu Sans", 26, bold=True)
        self._font_item = pygame.font.SysFont("DejaVu Sans", 20)
        self._font_sub  = pygame.font.SysFont("DejaVu Sans", 14)

    def on_enter(self) -> None:
        self.group_idx = 0
        self.ch_idx = 0
        self.pane = "groups"
        self._load_groups()

    def _load_groups(self) -> None:
        try:
            resp = self.shell.api.get("/iptv/groups")
            self.groups = resp.json() if resp.is_success else []
            if not self.groups:
                self.groups = ["All"]
        except Exception:
            self.groups = ["All"]
        self._load_channels()

    def _load_channels(self) -> None:
        group = self.groups[self.group_idx] if self.groups else None
        try:
            params = {} if group == "All" else {"group": group}
            resp = self.shell.api.get("/iptv/channels", params=params)
            self.channels = resp.json() if resp.is_success else []
        except Exception:
            self.channels = []
        self.ch_idx = 0

    def draw(self, surface: pygame.Surface) -> None:
        self._ensure_fonts()
        W, H = surface.get_size()

        # Header
        header = self._font_hdr.render("📡  IPTV", True, ACCENT)
        surface.blit(header, (MARGIN_X, 60))

        hint = self._font_sub.render(
            "← → switch pane  |  ↑ ↓ navigate  |  OK play  |  Back home",
            True, GRAY,
        )
        surface.blit(hint, (MARGIN_X, 100))

        # Separator
        pygame.draw.line(surface, (50, 50, 50),
                         (MARGIN_X + GROUP_W + 20, MARGIN_Y - 10),
                         (MARGIN_X + GROUP_W + 20, H - 40), 1)

        # Groups pane
        visible = (H - MARGIN_Y) // ITEM_H
        g_start = max(0, self.group_idx - visible // 2)
        for i, grp in enumerate(self.groups[g_start:g_start + visible]):
            idx = g_start + i
            y = MARGIN_Y + i * ITEM_H
            selected = (idx == self.group_idx)
            active_pane = (self.pane == "groups")

            if selected:
                bg = TILE_HL if active_pane else (40, 40, 50)
                pygame.draw.rect(surface, bg, (MARGIN_X, y, GROUP_W, ITEM_H - 4), border_radius=8)
                if active_pane:
                    pygame.draw.rect(surface, ACCENT, (MARGIN_X, y, GROUP_W, ITEM_H - 4), width=2, border_radius=8)

            color = WHITE if selected else GRAY
            surf = self._font_item.render(grp[:22], True, color)
            surface.blit(surf, (MARGIN_X + 12, y + 14))

        # Channels pane
        ch_x = MARGIN_X + GROUP_W + 40
        if not self.channels:
            msg = self._font_item.render("No channels", True, GRAY)
            surface.blit(msg, (ch_x, MARGIN_Y + 20))
        else:
            c_start = max(0, self.ch_idx - visible // 2)
            for i, ch in enumerate(self.channels[c_start:c_start + visible]):
                idx = c_start + i
                y = MARGIN_Y + i * ITEM_H
                selected = (idx == self.ch_idx)
                active_pane = (self.pane == "channels")

                if selected:
                    bg = TILE_HL if active_pane else (40, 40, 50)
                    pygame.draw.rect(surface, bg, (ch_x, y, LIST_W, ITEM_H - 4), border_radius=8)
                    if active_pane:
                        pygame.draw.rect(surface, ACCENT, (ch_x, y, LIST_W, ITEM_H - 4), width=2, border_radius=8)

                color = WHITE if selected else GRAY
                surf = self._font_item.render(ch.get("name", "")[:48], True, color)
                surface.blit(surf, (ch_x + 12, y + 14))

    def handle_key(self, key: str) -> None:
        if key == "right" and self.pane == "groups":
            self.pane = "channels"
        elif key == "left" and self.pane == "channels":
            self.pane = "groups"
        elif key == "up":
            if self.pane == "groups":
                self.group_idx = max(0, self.group_idx - 1)
                self._load_channels()
            else:
                self.ch_idx = max(0, self.ch_idx - 1)
        elif key == "down":
            if self.pane == "groups" and self.groups:
                self.group_idx = min(len(self.groups) - 1, self.group_idx + 1)
                self._load_channels()
            elif self.pane == "channels" and self.channels:
                self.ch_idx = min(len(self.channels) - 1, self.ch_idx + 1)
        elif key == "ok":
            if self.pane == "channels" and self.channels:
                ch = self.channels[self.ch_idx]
                try:
                    self.shell.api.post(f"/iptv/play/{ch['id']}")
                except Exception:
                    pass
        elif key in ("back", "home"):
            self.shell.navigate("home")

File: core/src/ui/iptv.c
#include "iptv.h"
#include "home.h"
#include "../render.h"
#include "../font.h"
#include "../iptv.h"
#include "../mpv.h"
#include "../history.h"
#include <string.h>
#include <stdio.h>

/* ── Layout constants (1920×1080 base) ─────────────────────────────────── */
#define HEADER_H   112   /* height of top header bar */
#define FOOTER_H    44   /* height of bottom hint bar */
#define MARGIN_X    48
#define GROUP_W    288   /* width of the left groups pane */
#define SEP_W        2   /* separator between panes */
#define SEP_X       (MARGIN_X + GROUP_W)
#define LIST_X      (SEP_X + SEP_W + 10)
#define ITEM_H      56   /* row height for both panes */
#define NUM_W       56   /* channel number column width */

typedef enum { PANE_GROUPS, PANE_CHANNELS } Pane;

static Pane         g_pane      = PANE_GROUPS;
static int          g_group_idx = 0;   /* 0 = "All channels" virtual group */
static int          g_ch_idx    = 0;

static const char **g_groups    = NULL;
static int          g_group_n   = 0;   /* count of real groups */

static IptvChannel *g_channels  = NULL;
static int          g_ch_n      = 0;

/* URL of the channel currently being played — for the "> playing" indicator */
static char g_playing_url[512] = "";

/* ── Helpers ─────────────────────────────────────────────────────────────── */

/* Total virtual group count including the "All channels" entry at index 0. */
static int total_groups(void) { return g_group_n + 1; }

/* Label for a virtual group index. */
static const char *group_label(int idx) {
    if (idx == 0) return "All channels";
    return g_groups[idx - 1];
}

static void load_channels(void) {
    const char *grp = (g_group_idx == 0) ? NULL : g_groups[g_group_idx - 1];
    g_channels = iptv_get_channels(NULL, grp, &g_ch_n);
    g_ch_idx   = 0;
}

/* ── Public API ──────────────────────────────────────────────────────────── */

void ui_iptv_enter(void) {
    g_pane      = PANE_GROUPS;
    g_group_idx = 0;
    g_groups    = iptv_get_groups(&g_group_n);
    load_channels();
}

void ui_iptv_draw(void) {
    int W = g_screen_w;
    int H = g_screen_h;
    int list_w    = W - LIST_X - MARGIN_X;
    int content_h = H - HEADER_H - FOOTER_H;
    int visible   = content_h / ITEM_H;

    /* ── Header ────────────────────────────────────────────────────────── */
    font_draw(MARGIN_X, 24, "IPTV", 54, COL_ACCENT);

    /* Group name + channel count on the right side of header */
    char hdr[96];
    snprintf(hdr, sizeof(hdr), "%s  /  %d", group_label(g_group_idx), g_ch_n);
    font_draw(LIST_X, 38, hdr, 24, COL_GRAY);

    /* Thin separator below header */
    render_rect(0, HEADER_H, W, 1, rgba(50, 50, 70, 255));

    /* ── Groups pane ───────────────────────────────────────────────────── */
    int g_total = total_groups();
    int g_start = g_group_idx - visible / 2;
    if (g_start < 0) g_start = 0;

    for (int i = 0; i < visible && (g_start + i) < g_total; i++) {
        int idx = g_start + i;
        int y   = HEADER_H + i * ITEM_H;
        int sel = (idx == g_group_idx);
        int act = sel && (g_pane == PANE_GROUPS);

        if (sel) {
            uint32_t bg = act ? COL_ACCENT : rgba(38, 38, 58, 255);
            render_rect(MARGIN_X + 4, y + 5, GROUP_W - 8, ITEM_H - 10, bg);
            if (act)
                render_rect_outline(MARGIN_X + 4, y + 5,
                                    GROUP_W - 8, ITEM_H - 10, COL_ACCENT, 2);
        }

        char label[34] = {0};
        strncpy(label, group_label(idx), 32);
        uint32_t col = act  ? COL_WHITE :
                       sel  ? rgba(180, 195, 255, 255) :
                               COL_GRAY;
        font_draw(MARGIN_X + 14, y + ITEM_H / 2 - 12, label, 22, col);
    }

    /* Groups scroll indicator (thin bar on the far left) */
    if (g_total > visible) {
        int bar_h = content_h * visible / g_total;
        int bar_y = HEADER_H + content_h * g_start / g_total;
        render_rect(MARGIN_X, HEADER_H, 3, content_h, rgba(30, 30, 45, 255));
        render_rect(MARGIN_X, bar_y, 3, bar_h, rgba(80, 90, 160, 255));
    }

    /* Vertical separator between panes */
    render_rect(SEP_X, HEADER_H, SEP_W, content_h, rgba(45, 45, 65, 255));

    /* ── Channels pane ─────────────────────────────────────────────────── */
    if (g_ch_n == 0) {
        font_draw(LIST_X + 20, HEADER_H + 40, "No channels", 24, COL_GRAY);
    } else {
        int c_start = g_ch_idx - visible / 2;
        if (c_start < 0) c_start = 0;

        for (int i = 0; i < visible && (c_start + i) < g_ch_n; i++) {
            int idx     = c_start + i;
            int y       = HEADER_H + i * ITEM_H;
            int sel     = (idx == g_ch_idx);
            int act     = sel && (g_pane == PANE_CHANNELS);
            int playing = g_playing_url[0] &&
                          !strcmp(g_channels[idx].url, g_playing_url);

            /* Row background */
            if (act) {
                render_rect(LIST_X, y + 4, list_w, ITEM_H - 8, COL_TILE_HL);
                render_rect_outline(LIST_X, y + 4, list_w, ITEM_H - 8,
                                    COL_ACCENT, 2);
            } else if (sel) {
                render_rect(LIST_X, y + 4, list_w, ITEM_H - 8,
                            rgba(38, 38, 56, 255));
            } else if (playing) {
                render_rect(LIST_X, y + 4, list_w, ITEM_H - 8,
                            rgba(18, 42, 18, 255));
            }

            /* Channel number */
            char num[8];
            snprintf(num, sizeof(num), "%d", idx + 1);
            font_draw(LIST_X + 8, y + ITEM_H / 2 - 11, num, 19,
                      act ? COL_ACCENT : rgba(90, 90, 115, 255));

            /* Channel name */
            char name[54] = {0};
            strncpy(name, g_channels[idx].name, 52);
            uint32_t col = act     ? COL_WHITE  :
                           sel     ? rgba(220, 225, 255, 255) :
                           playing ? rgba(110, 210, 110, 255) :
                                     rgba(185, 185, 200, 255);
            font_draw(LIST_X + NUM_W, y + ITEM_H / 2 - 11, name, 24, col);

            /* Playing indicator */
            if (playing) {
                float pw = font_measure(">", 22);
                font_draw(LIST_X + list_w - (int)pw - 12,
                          y + ITEM_H / 2 - 11, ">", 22,
                          rgba(90, 210, 90, 255));
            }
        }

        /* Channels scroll bar (right edge) */
        if (g_ch_n > visible) {
            int bar_h = content_h * visible / g_ch_n;
            int bar_y = HEADER_H + content_h * c_start / g_ch_n;
            render_rect(W - MARGIN_X + 2, HEADER_H,
                        4, content_h, rgba(28, 28, 42, 255));
            render_rect(W - MARGIN_X + 2, bar_y,
                        4, bar_h,     rgba(80, 90, 160, 255));
        }
    }

    /* ── Footer ────────────────────────────────────────────────────────── */
    render_rect(0, H - FOOTER_H, W, 1, rgba(50, 50, 70, 255));

    const char *hint = (g_pane == PANE_GROUPS)
        ? "up/down: group   right: channels   back: home"
        : "up/down: channel   left: groups   ok: play   back: home";
    font_draw(MARGIN_X, H - FOOTER_H + 13, hint, 19, COL_GRAY);

    /* Now-playing name (right side of footer) */
    if (g_playing_url[0]) {
        for (int i = 0; i < g_ch_n; i++) {
            if (!strcmp(g_channels[i].url, g_playing_url)) {
                char np[80];
                snprintf(np, sizeof(np), "> %.50s", g_channels[i].name);
                float tw = font_measure(np, 19);
                font_draw((int)(W - tw - MARGIN_X), H - FOOTER_H + 13,
                          np, 19, rgba(90, 210, 90, 255));
                break;
            }
        }
    }
}

void ui_iptv_key(const char *key) {
    if (!strcmp(key, "right") && g_pane == PANE_GROUPS) {
        g_pane = PANE_CHANNELS;

    } else if (!strcmp(key, "left") && g_pane == PANE_CHANNELS) {
        g_pane = PANE_GROUPS;

    } else if (!strcmp(key, "up")) {
        if (g_pane == PANE_GROUPS && g_group_idx > 0) {
            g_group_idx--;
            load_channels();
        } else if (g_pane == PANE_CHANNELS && g_ch_idx > 0) {
            g_ch_idx--;
        }

    } else if (!strcmp(key, "down")) {
        if (g_pane == PANE_GROUPS && g_group_idx < total_groups() - 1) {
            g_group_idx++;
            load_channels();
        } else if (g_pane == PANE_CHANNELS &&
                   g_channels && g_ch_idx < g_ch_n - 1) {
            g_ch_idx++;
        }

    } else if (!strcmp(key, "ok")) {
        if (g_pane == PANE_CHANNELS && g_channels && g_ch_n > 0) {
            IptvChannel *ch = &g_channels[g_ch_idx];
            strncpy(g_playing_url, ch->url, sizeof(g_playing_url) - 1);
            mpv_core_load(ch->url, "live");
            history_record(ch->url, ch->name, "iptv", ch->name, ch->logo, 0);
        } else if (g_pane == PANE_GROUPS) {
            /* Enter channels pane on OK from groups */
            g_pane = PANE_CHANNELS;
        }

    } else if (!strcmp(key, "back") || !strcmp(key, "home")) {
        /* Stop playback before going home so the home screen renders */
        mpv_core_stop();
        navigate("home");
    }
}
